ENG-801 Compile database before dev

maparent opened the pull request:
https://linear.app/discourse-graphs/issue/ENG-801/compile-database-before-dev
Took on much of the scope of [ENG-799](https://linear.app/discourse-graphs/issue/ENG-799/distinguish-compilation-from-code-generation-in-database)

*    Precompile dbDotEnv so database does not have to be built beforehand
*    Introduce SUPABASE_USE_DB=none. (Now the default value.)
*    dev and build in database will not activate supabase if that is set.


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

- New Features
  - Added support to run the app without a database (now the default), automatically skipping local DB setup when not needed.
- Documentation
  - Updated usage scenarios to include “Developing without the database.”
  - Clarified local development steps, including how to opt into a local database.
- Chores
  - Streamlined build and development workflows for database-related tasks.
  - Separated schema generation into its own task for clearer, faster builds.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

maparent committed:
In @repo/database, separate tsc from building the schema.
Make build depend on building the schema, but not check-types.
(Still thinking about the pros and cons here)

maparent committed:
Compile database before running dev

mdroidian reviewed (changes requested):
Does this suggest we need to have docker/supabase running to run roam dev?

We want to try and avoid this if at all possible.

maparent committed:
Precompile dbDotEnv so database does not have to be built
Introduce SUPABASE_USE_DB=none. (Now the default value.)
dev and build in database will not activate supabase if that is set.

maparent commented on packages/database/scripts/build.ts:
```diff
@@ -1,13 +1,17 @@
 
 if (process.env.HOME !== "/vercel") {
   try {
+    if (getVariant() === "none") {
+      console.log("Not using the database");
+      process.exit(0);
+    }
```
Actually we will be using the database in the null case. This could be clearer, but such clarifications will be in a later PR.

maparent commented on packages/database/scripts/dev.ts:
```diff
@@ -0,0 +1,20 @@
+import { execSync } from "node:child_process";
+import { join, dirname } from "path";
+import { getVariant } from "@repo/database/dbDotEnv";
+
+const __dirname = dirname(__filename);
+const projectRoot = join(__dirname, "..");
+
```
This seems to make it less likely to run as a cjs import. It is more important for it to run under cjs than ejs.

maparent commented on packages/database/scripts/dev.ts:
```diff
@@ -0,0 +1,20 @@
+if (process.env.HOME !== "/vercel") {
+  try {
+    execSync("npm run compile", { cwd: projectRoot, stdio: "inherit" });
+    if (getVariant() === "none") {
+      console.log("Not using the database");
+      process.exit(0);
+    }
```
Again, (confusing) wrong assumption.

maparent replied:
Renamed null to "implicit"

maparent commented on packages/database/src/dbDotEnv.js:
```diff
@@ -1,34 +1,33 @@
-import { fileURLToPath } from "url";
 
-const findRoot = (): string => {
+const findRoot = () => {
   let dir = __filename;
   while (basename(dir) !== "database") {
     dir = dirname(dir);
   }
   return dir;
 };
```
Again, this would fail in cjs. The infinite loop is theoretical, as this code exists in the database folder.

maparent commented on packages/database/src/dbDotEnv.js:
```diff
@@ -1,34 +1,33 @@
     throw new Error("Invalid variant: " + variant);
   }
   return variant;
 };
```
using implicit vs null, should be clearer.

maparent committed:
coderabbit suggestions
